### DISCUSSION TITLE
How to define custom SPARQL extension functions (v3.x)

### DISCUSSION THREAD

**Discussion by ghost** [dotnetrdf/dotnetrdf#598] [Q&A]
Hello,
I have upgraded from v2.x to v3.x as I need rdf-star and sparql-star support.
I would like to define some custom sparql extension functions. While I did this quite easily in v2.x using the documentation here: https://github.com/dotnetrdf/dotnetrdf/blob/main/docs/developer_guide/sparql/extension_functions.md, it seems that this way of implementing custom extension functions is no longer compatible with v3.0.

Would you please give me some insight on how to implement this in v3.x?

Thank you very much!

**Comment by kal** [comment]
Hi @grocher06 

I'm afraid that this was an undocumented breaking change in 3.0. I'm really sorry about doing that! The refactoring was done in preparation for implementing a second, streaming query engine. But in doing that I think that basically I broke the API quite badly. The changes have made it much harder to implement extension functions, and it also means that the processor ends up being a massive monolithic class with all the implementations contained in it. Essentially with 3.0 to add extensions you would need to both add the exoression class and extended LeviathanExpressionProcessor to add methods to handle processing the expression.

I feel like I need to find a way to restructure the code back to the old approach where each expression implementation provides either a block processor implementation (for Leviathan) or a streaming processor implementation (for the new engine) or both. The streaming engine work is at very early stages right now so really I shouldn't have introduced these changes into the 3.0 release at all :-(

**Reply by ghost** [reply]
Thanks for your feedback. So I'll have to live with some hacks and wait for the feature to get back into shape ;-)
Anyway, thanks for your great work, this lib rocks!
Regards.
